### PR TITLE
topdown: Add built-in HMAC functions

### PR DESCRIPTION
**Pull request by johanfylling**
Add crypto.hmac.* built-in functions for the MD5, SHA-1, SHA-256 and SHA-512 hashing algorithms.

Add documentation for how to contribute new built-in functions.

Fixes: #1740
Signed-off-by: Johan Fylling <johan.dev@fylling.se>

### CODE REVIEW

File: docs/content/contrib-adding-builtin-functions.md
Comment (johanfylling) [inline docs/content/contrib-adding-builtin-functions.md:1]: I just realized there is more in-depth documentation for built-in:s at `extensions/#custom-built-in-functions-in-go`. Should probably link to there from here.

File: docs/README.md
Comment (srenatus) [inline docs/README.md:50]: [nit] Isn't it "IDEs" rather than "IDE's"? Just since your touching the line anyways 🤔

File: docs/content/contrib-adding-builtin-functions.md
Comment (srenatus) [inline docs/content/contrib-adding-builtin-functions.md:154]: Haven't seen any wasm mention here, so I think this should say "SDK dependent"

**Review by srenatus — COMMENTED** [review]
Two nitpicks from skimming the PR. I'll have another look tomorrow.

File: test/cases/testdata/cryptohmacsha256/test-cryptohmacsha256.yaml
Comment (anderseknert) [inline test/cases/testdata/cryptohmacsha256/test-cryptohmacsha256.yaml:12]: It would be good with at least one test providing unicode characters in string and/or key.

File: docs/content/contrib-adding-builtin-functions.md
Comment (srenatus) [reply docs/content/contrib-adding-builtin-functions.md:1]: That is a little different, though: it's about extending OPA, so you'd build your own binary with the extra built-in func included. This document describes how to contribute to OPA, so the new built-in function would become part of "core OPA" so to speak. There certainly is overlap, but it's a different twist.

File: docs/content/contrib-adding-builtin-functions.md
Comment (srenatus) [inline docs/content/contrib-adding-builtin-functions.md:67]: [nit] Let's make it three blocks: stdlib, 3rd party, internal:
```suggestion
import (
    "strings"
    
    "github.com/pkg/errors"
    
    "github.com/open-policy-agent/opa/ast"
    "github.com/open-policy-agent/opa/topdown/builtins"
)
```

and let's not provide examples using pkg/errors: https://github.com/open-policy-agent/opa/issues/2152

**Comment by tsandall** [comment]
LGTM modulo the comments that @srenatus and @anderseknert posted.

File: docs/content/contrib-adding-builtin-functions.md
Comment (johanfylling) [reply docs/content/contrib-adding-builtin-functions.md:154]: Changing to `SDK-dependent`. Would probably be good to also add info to this how-to about Wasm. I don't think I can currently bild and test for that on my M1 mac; so something to return to later, perhaps, either for me or someone else.

Comment (srenatus) [reply docs/content/contrib-adding-builtin-functions.md:154]: It's cumbersome, but you could perhaps use a docker container...? Not sure how that's done on M1.

File: docs/content/contrib-adding-builtin-functions.md
Comment (anderseknert) [reply docs/content/contrib-adding-builtin-functions.md:1]: It is different for sure, but I think @johanfylling is right that it would be good to reference here.. at least in the sense that one step that's currently not mentioned is to first have some form of approval for the built-in. Without that—i.e. for someone who wants to integrate Kafka/Redis/DynamoDB/Whatever—the custom function is probably a better alternative.

File: test/cases/testdata/cryptohmacsha256/test-cryptohmacsha256.yaml
Comment (johanfylling) [reply test/cases/testdata/cryptohmacsha256/test-cryptohmacsha256.yaml:12]: Looking into this.

File: docs/content/contrib-adding-builtin-functions.md
Comment (srenatus) [reply docs/content/contrib-adding-builtin-functions.md:1]: On a related note, this feels a bit like these are two changes: a docs addition and new built-in functions.

It would be nice if we untangled that a bit, we could massage it into two commits -- two PRs would be another option, but since it's here already, it's OK I think.

Comment (johanfylling) [reply docs/content/contrib-adding-builtin-functions.md:1]: Yea, I agree; I should have gone that route.

File: topdown/crypto.go
Comment (srenatus) [inline topdown/crypto.go:196]: 💭  Why `h func() hash.Hash` and not `h hash.Hash`? 💡 It's the `hmac.New()` interface that makes this the right choice. Interesting.

File: test/cases/testdata/cryptohmacsha256/test-cryptohmacsha256.yaml
Comment (anderseknert) [reply test/cases/testdata/cryptohmacsha256/test-cryptohmacsha256.yaml]: New tests are great!

**Comment by tsandall** [comment]
This looks like it's ready to go to me. Let's deal with the Wasm changes separately. As soon as this is squashed, we can merge.

**Review by srenatus — APPROVED** [review]

**Pull request merged**
